File: wallet/presenter/plugins/oid4vp/response_encryption.go
package oid4vp

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/go-jose/go-jose/v4"
	"github.com/trustknots/vcknots/wallet/common/hpke"
	"github.com/trustknots/vcknots/wallet/common/jwks"
	"github.com/trustknots/vcknots/wallet/presenter/types"
)

// defaultResponseContentEncryption is the JWE "enc" value OID4VP 1.1 Section 8.3
// falls back to when the Verifier does not publish
// encrypted_response_enc_values_supported. It does not apply to JOSE HPKE, where
// no separate content encryption algorithm exists.
const defaultResponseContentEncryption = "A128GCM"

// sessionInfoPrefixRedirect and sessionInfoPrefixDCAPI are the fixed labels that
// open the session_info structure of OID4VP 1.1 Section 8.3.1. Which one applies
// depends on how the presentation was invoked.
const (
	sessionInfoPrefixRedirect = "OpenID4VP-si"
	sessionInfoPrefixDCAPI    = "OpenID4VPDCAPI-si"
)

// sessionInfoSeparator is the 0xFF byte that delimits the session_info fields.
const sessionInfoSeparator = 0xFF

// RedirectSessionInfo builds the session_info structure for the response modes
// that are invoked through redirects, direct_post.jwt among them:
//
//	session_info = ASCII("OpenID4VP-si") || BYTE(255) || ASCII(clientId) ||
//	               BYTE(255) || ASCII(nonce) || BYTE(255) || ASCII(responseUri)
//
// clientID is the client_id request parameter including its Client Identifier
// Prefix, and responseURI is whichever of response_uri or redirect_uri the
// Response Mode uses.
//
// The Wallet and the Verifier compute this independently and feed it to HPKE as
// the info parameter, so a response encrypted for one session cannot be
// decrypted in the context of another: decryption fails closed rather than
// yielding a credential that was bound to different session data.
func RedirectSessionInfo(clientID, nonce, responseURI string) []byte {
	return buildSessionInfo(sessionInfoPrefixRedirect, clientID, nonce, responseURI)
}

// DCAPISessionInfo builds the session_info structure for the dc_api.jwt Response
// Mode:
//
//	session_info = ASCII("OpenID4VPDCAPI-si") || BYTE(255) || ASCII(origin) ||
//	               BYTE(255) || ASCII(nonce)
//
// origin is the Origin of the request and must not carry the "origin:" prefix.
func DCAPISessionInfo(origin, nonce string) []byte {
	return buildSessionInfo(sessionInfoPrefixDCAPI, origin, nonce)
}

func buildSessionInfo(prefix string, fields ...string) []byte {
	info := []byte(prefix)
	for _, field := range fields {
		info = append(info, sessionInfoSeparator)
		info = append(info, field...)
	}
	return info
}

// sessionInfoForRequest returns the session_info structure that matches the
// Response Mode of the request.
func sessionInfoForRequest(req *types.PresentationRequest) []byte {
	if req == nil {
		return nil
	}

	if OAuthAuthzReqResponseMode(req.ResponseMode) == OAuthAuthzReqResponseModeDCAPIJWT {
		// The Digital Credentials API identifies the session by the Origin of
		// the request rather than by a response endpoint.
		return DCAPISessionInfo(req.Origin, req.Nonce)
	}
	return RedirectSessionInfo(req.ClientID, req.Nonce, req.ResponseURI)
}

// responseEncryptionKey is the Verifier public key an Authorization Response is
// encrypted to, together with the JWE "alg" that the key pins.
type responseEncryptionKey struct {
	key jwks.Key
	alg string
}

// selectResponseEncryptionKey picks the key to encrypt the Authorization
// Response to, following OID4VP 1.1 Section 8.3: keys whose "use" is absent or
// "enc" are candidates, the "alg" member must be present, and the JWE "alg" is
// the one the key declares.
//
// When the Verifier offers several usable keys the JOSE HPKE ones win. HPKE
// binds the response to the session through the info parameter of Section
// 8.3.1, which the ECDH-ES algorithms have no equivalent for.
func selectResponseEncryptionKey(set jwks.Set, preferredAlg string) (responseEncryptionKey, error) {
	var hpkeCandidates, joseCandidates []responseEncryptionKey
	var skipped []string

	for _, key := range set.Keys {
		if key.Use != "" && key.Use != "enc" {
			skipped = append(skipped, fmt.Sprintf("%q has use %q", key.KeyID, key.Use))
			continue
		}
		if key.Algorithm == "" {
			skipped = append(skipped, fmt.Sprintf("%q declares no alg", key.KeyID))
			continue
		}

		candidate := responseEncryptionKey{key: key, alg: key.Algorithm}
		switch {
		case hpke.Algorithm(key.Algorithm).Supported():
			hpkeCandidates = append(hpkeCandidates, candidate)
		case hpke.IsAlgorithm(key.Algorithm):
			// A JOSE HPKE algorithm this build cannot perform, such as one of
			// the X448 suites or a Key Encryption variant.
			skipped = append(skipped, fmt.Sprintf("%q uses unsupported JOSE HPKE alg %q", key.KeyID, key.Algorithm))
		case isSupportedJWEKeyAlgorithm(key.Algorithm):
			joseCandidates = append(joseCandidates, candidate)
		default:
			skipped = append(skipped, fmt.Sprintf("%q uses unsupported alg %q", key.KeyID, key.Algorithm))
		}
	}

	candidates := append(append([]responseEncryptionKey{}, hpkeCandidates...), joseCandidates...)
	if len(candidates) == 0 {
		if len(skipped) > 0 {
			return responseEncryptionKey{}, fmt.Errorf("no usable response encryption key in the verifier jwks: %s", strings.Join(skipped, "; "))
		}
		return responseEncryptionKey{}, fmt.Errorf("the verifier jwks holds no response encryption key")
	}

	// An explicitly requested algorithm, which is how the draft-24
	// authorization_encrypted_response_alg metadata reaches this code, narrows
	// the choice rather than replacing it.
	if preferredAlg != "" {
		for _, candidate := range candidates {
			if candidate.alg == preferredAlg {
				return candidate, nil
			}
		}
	}

	// Preferring HPKE over ECDH-ES also means preferring the suite the wallet
	// lists first among the HPKE algorithms it implements.
	if len(hpkeCandidates) > 1 {
		for _, alg := range hpke.SupportedAlgorithms {
			for _, candidate := range hpkeCandidates {
				if candidate.alg == string(alg) {
					return candidate, nil
				}
			}
		}
	}

	return candidates[0], nil
}

// encryptAuthorizationResponse encrypts the Authorization Response parameters
// and returns the JWE Compact Serialization that goes into the "response"
// parameter (OID4VP 1.1 Section 8.3).
//
// sessionInfo is the session_info structure of Section 8.3.1. It is used only by
// the JOSE HPKE algorithms; the ECDH-ES algorithms have nowhere to put it.
func encryptAuthorizationResponse(payload map[string]any, metadata *VerifierMetadata, sessionInfo []byte) (string, error) {
	if metadata == nil {
		return "", fmt.Errorf("verifier metadata is not available for response encryption")
	}
	if len(metadata.Jwks.Keys) == 0 {
		return "", fmt.Errorf("verifier jwks is not available for response encryption")
	}

	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to marshal the authorization response: %w", err)
	}

	selected, err := selectResponseEncryptionKey(metadata.Jwks, metadata.AuthorizationEncryptedResponseAlg)
	if err != nil {
		return "", err
	}

	if alg := hpke.Algorithm(selected.alg); alg.Supported() {
		return encryptWithHPKE(alg, selected.key, payloadBytes, sessionInfo)
	}
	return encryptWithJWEKeyAgreement(selected, metadata.contentEncryptionAlgorithm(), payloadBytes)
}

// encryptWithHPKE produces the response JWE using JOSE HPKE Integrated
// Encryption, which is the mode OID4VP 1.1 Section 8.3.1 builds on.
func encryptWithHPKE(alg hpke.Algorithm, key jwks.Key, payload, sessionInfo []byte) (string, error) {
	publicKey, err := key.ECDHPublicKey()
	if err != nil {
		return "", fmt.Errorf("verifier encryption key %q is unusable: %w", key.KeyID, err)
	}

	header := map[string]any{}
	// Section 8.3 requires echoing the kid so the Verifier can tell which of its
	// keys the response was encrypted to.
	if key.KeyID != "" {
		header["kid"] = key.KeyID
	}

	compact, err := hpke.EncryptJWE(alg, publicKey, header, sessionInfo, payload)
	if err != nil {
		return "", fmt.Errorf("failed to encrypt the authorization response with %s: %w", alg, err)
	}
	return compact, nil
}

// encryptWithJWEKeyAgreement produces the response JWE using the ECDH-ES family,
// which OID4VP has supported since before JOSE HPKE existed.
func encryptWithJWEKeyAgreement(selected responseEncryptionKey, contentEnc string, payload []byte) (string, error) {
	keyAlg, err := parseJWEKeyAlgorithm(selected.alg)
	if err != nil {
		return "", err
	}
	contentEncryption, err := parseJWEContentEncryption(contentEnc)
	if err != nil {
		return "", err
	}

	joseKey, err := selected.key.JOSE()
	if err != nil {
		return "", fmt.Errorf("verifier encryption key %q is unusable: %w", selected.key.KeyID, err)
	}

	encrypter, err := jose.NewEncrypter(
		contentEncryption,
		jose.Recipient{
			Algorithm: keyAlg,
			Key:       joseKey.Key,
			KeyID:     joseKey.KeyID,
		},
		nil,
	)
	if err != nil {
		return "", fmt.Errorf("failed to create encrypter: %w", err)
	}

	jwe, err := encrypter.Encrypt(payload)
	if err != nil {
		return "", fmt.Errorf("failed to encrypt the authorization response: %w", err)
	}

	serialized, err := jwe.CompactSerialize()
	if err != nil {
		return "", fmt.Errorf("failed to serialize JWE: %w", err)
	}
	return serialized, nil
}

// contentEncryptionAlgorithm returns the JWE "enc" value to use with the
// ECDH-ES algorithms, preferring the OID4VP 1.1 metadata over its draft-24
// predecessor and falling back to the default the specification allows.
func (v *VerifierMetadata) contentEncryptionAlgorithm() string {
	for _, enc := range v.EncryptedResponseEncValuesSupported {
		if _, err := parseJWEContentEncryption(enc); err == nil {
			return enc
		}
	}
	if v.AuthorizationEncryptedResponseEnc != "" {
		return v.AuthorizationEncryptedResponseEnc
	}
	return defaultResponseContentEncryption
}

func isSupportedJWEKeyAlgorithm(alg string) bool {
	_, err := parseJWEKeyAlgorithm(alg)
	return err == nil
}

func parseJWEKeyAlgorithm(alg string) (jose.KeyAlgorithm, error) {
	switch alg {
	case "ECDH-ES":
		return jose.ECDH_ES, nil
	case "ECDH-ES+A128KW":
		return jose.ECDH_ES_A128KW, nil
	case "ECDH-ES+A192KW":
		return jose.ECDH_ES_A192KW, nil
	case "ECDH-ES+A256KW":
		return jose.ECDH_ES_A256KW, nil
	default:
		return "", fmt.Errorf("unsupported encryption algorithm: %s", alg)
	}
}

func parseJWEContentEncryption(enc string) (jose.ContentEncryption, error) {
	switch enc {
	case "A128GCM":
		return jose.A128GCM, nil
	case "A192GCM":
		return jose.A192GCM, nil
	case "A256GCM":
		return jose.A256GCM, nil
	case "A128CBC-HS256":
		return jose.A128CBC_HS256, nil
	case "A192CBC-HS384":
		return jose.A192CBC_HS384, nil
	case "A256CBC-HS512":
		return jose.A256CBC_HS512, nil
	default:
		return "", fmt.Errorf("unsupported encryption encoding: %s", enc)
	}
}

File: issuer+verifier/src/response-encryption.ts
import { compactDecrypt, importJWK } from 'jose'
import { err } from './errors/vcknots.error'
import { decryptHpkeJwe, isHpkeAlgorithm, isSupportedHpkeAlgorithm } from './jose-hpke'
import { Jwk } from './jwk.type'

/**
 * Encrypted Authorization Responses, OID4VP 1.1 Section 8.3.
 *
 * The Verifier publishes encryption keys in `client_metadata.jwks`, asks for a
 * `.jwt` Response Mode, and receives the Authorization Response as a single
 * `response` parameter holding an unsigned, encrypted JWT.
 */

/**
 * The fixed ASCII labels that open the session_info structure of Section 8.3.1.
 * Which one applies depends on how the presentation was invoked.
 */
const SESSION_INFO_PREFIX_REDIRECT = 'OpenID4VP-si'
const SESSION_INFO_PREFIX_DC_API = 'OpenID4VPDCAPI-si'

/** The 0xFF byte that delimits the session_info fields. */
const SESSION_INFO_SEPARATOR = 0xff

const buildSessionInfo = (prefix: string, ...fields: string[]): Buffer =>
  Buffer.concat([
    Buffer.from(prefix, 'ascii'),
    ...fields.flatMap((field) => [
      Buffer.from([SESSION_INFO_SEPARATOR]),
      Buffer.from(field, 'ascii'),
    ]),
  ])

/**
 * The session_info structure for the Response Modes invoked through redirects,
 * `direct_post.jwt` among them:
 *
 * ```
 * session_info = ASCII("OpenID4VP-si") || BYTE(255) || ASCII(clientId) ||
 *                BYTE(255) || ASCII(nonce) || BYTE(255) || ASCII(responseUri)
 * ```
 *
 * `clientId` is the client_id request parameter including its Client Identifier
 * Prefix, and `responseUri` is whichever of response_uri or redirect_uri the
 * Response Mode uses.
 */
export const redirectSessionInfo = (clientId: string, nonce: string, responseUri: string): Buffer =>
  buildSessionInfo(SESSION_INFO_PREFIX_REDIRECT, clientId, nonce, responseUri)

/**
 * The session_info structure for the `dc_api.jwt` Response Mode:
 *
 * ```
 * session_info = ASCII("OpenID4VPDCAPI-si") || BYTE(255) || ASCII(origin) ||
 *                BYTE(255) || ASCII(nonce)
 * ```
 *
 * `origin` is the Origin of the request and must not carry the `origin:` prefix.
 */
export const dcApiSessionInfo = (origin: string, nonce: string): Buffer =>
  buildSessionInfo(SESSION_INFO_PREFIX_DC_API, origin, nonce)

/**
 * The session an encrypted Authorization Response belongs to.
 *
 * The Verifier reconstructs this from the request parameters it issued rather
 * than from anything the Wallet sends back, which is what makes HPKE decryption
 * fail closed on a response captured from a different session.
 */
export type ResponseEncryptionSession =
  | {
      responseMode: 'direct_post.jwt'
      /** The client_id request parameter, including its Client Identifier Prefix. */
      clientId: string
      nonce: string
      /** The response_uri, or the redirect_uri when the Response Mode uses one. */
      responseUri: string
    }
  | {
      responseMode: 'dc_api.jwt'
      /** The Origin of the request, without the `origin:` prefix. */
      origin: string
      nonce: string
    }

/** The session_info structure that matches a session's Response Mode. */
export const sessionInfoFor = (session: ResponseEncryptionSession): Buffer =>
  session.responseMode === 'dc_api.jwt'
    ? dcApiSessionInfo(session.origin, session.nonce)
    : redirectSessionInfo(session.clientId, session.nonce, session.responseUri)

/** The JWE `alg` values usable for response encryption alongside JOSE HPKE. */
const supportedKeyAgreementAlgorithms = [
  'ECDH-ES',
  'ECDH-ES+A128KW',
  'ECDH-ES+A192KW',
  'ECDH-ES+A256KW',
]

/** Whether an `alg` value can encrypt an Authorization Response to this library. */
export const isSupportedResponseEncryptionAlgorithm = (alg: unknown): alg is string =>
  isSupportedHpkeAlgorithm(alg) ||
  (typeof alg === 'string' && supportedKeyAgreementAlgorithms.includes(alg))

/**
 * Decrypts the `response` parameter of an encrypted Authorization Response and
 * returns its payload, which carries the response parameters as top-level JSON
 * members.
 *
 * `privateKeys` are the Verifier's response encryption keys. The `kid` header
 * parameter selects among them when present, as Section 8.3 requires the Wallet
 * to echo it; otherwise every key whose algorithm matches is tried.
 */
export const decryptAuthorizationResponse = async (
  response: string,
  privateKeys: Jwk[],
  session: ResponseEncryptionSession
): Promise<Record<string, unknown>> => {
  if (typeof response !== 'string' || response.trim() === '') {
    throw err('invalid_request', { message: 'The response parameter is empty.' })
  }
  if (privateKeys.length === 0) {
    throw err('authz_verifier_key_not_found', {
      message: 'The verifier holds no response encryption key.',
    })
  }

  const header = readProtectedHeader(response)
  const alg = header.alg
  if (!isSupportedResponseEncryptionAlgorithm(alg)) {
    throw err('invalid_encryption_parameters', {
      message: isHpkeAlgorithm(alg)
        ? `The response uses JOSE HPKE algorithm "${alg}", which this library does not implement.`
        : `The response uses unsupported encryption algorithm "${String(alg)}".`,
    })
  }

  const candidates = selectDecryptionKeys(privateKeys, alg, header.kid)
  if (candidates.length === 0) {
    throw err('authz_verifier_key_not_found', {
      message: header.kid
        ? `No verifier response encryption key matches kid "${header.kid}" and alg "${alg}".`
        : `No verifier response encryption key matches alg "${alg}".`,
    })
  }

  const sessionInfo = sessionInfoFor(session)
  const failures: string[] = []

  for (const key of candidates) {
    try {
      const payload = isSupportedHpkeAlgorithm(alg)
        ? decryptHpkeJwe(response, key, sessionInfo).plaintext
        : (await compactDecrypt(response, await importPrivateKey(key, alg))).plaintext
      return parsePayload(payload)
    } catch (error) {
      failures.push(`${key.kid ?? '(no kid)'}: ${error instanceof Error ? error.message : error}`)
    }
  }

  throw err('invalid_encryption_parameters', {
    message: `The authorization response could not be decrypted. ${failures.join('; ')}`,
  })
}

/**
 * Reads the `alg` and `kid` of a JWE compact serialization without decrypting,
 * so that the recipient can look up the key before it has one to decrypt with.
 */
const readProtectedHeader = (compact: string): { alg?: unknown; kid?: string } => {
  const parts = compact.split('.')
  if (parts.length !== 5) {
    throw err('invalid_encryption_parameters', {
      message: `Expected 5 JWE compact serialization parts, got ${parts.length}.`,
    })
  }
  try {
    const header = JSON.parse(Buffer.from(parts[0], 'base64url').toString('utf8'))
    return {
      alg: header?.alg,
      kid: typeof header?.kid === 'string' ? header.kid : undefined,
    }
  } catch (_error) {
    throw err('invalid_encryption_parameters', {
      message: 'The JWE protected header is not valid JSON.',
    })
  }
}

/**
 * Narrows the Verifier's keys to those that could have encrypted this response.
 * A key that pins a different `alg` is meant for a different algorithm, so it is
 * skipped rather than tried and reported as a decryption failure.
 */
const selectDecryptionKeys = (privateKeys: Jwk[], alg: string, kid?: string): Jwk[] =>
  privateKeys.filter((key) => {
    if (kid !== undefined && key.kid !== kid) return false
    return typeof key.alg !== 'string' || key.alg === alg
  })

const importPrivateKey = async (key: Jwk, alg: string) => {
  const imported = await importJWK(key as Parameters<typeof importJWK>[0], alg)
  if (imported instanceof Uint8Array) {
    throw err('invalid_encryption_parameters', {
      message: 'A symmetric key cannot decrypt an authorization response.',
    })
  }
  return imported
}

const parsePayload = (payload: Uint8Array): Record<string, unknown> => {
  let parsed: unknown
  try {
    parsed = JSON.parse(Buffer.from(payload).toString('utf8'))
  } catch (_error) {
    throw err('invalid_request', {
      message: 'The decrypted authorization response is not valid JSON.',
    })
  }
  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    throw err('invalid_request', {
      message: 'The decrypted authorization response is not a JSON object.',
    })
  }
  return parsed as Record<string, unknown>
}
